backpack/core/derivatives/basederivatives.py

```diff
@@ -124,17 +128,36 @@ def ea_jac_t_mat_jac_prod(self, module, g_inp, g_out, mat):
         """
         raise NotImplementedError
 
-    def hessian_is_zero(self):
+    def hessian_is_zero(self) -> bool:
+        """Whether ``∂²output[i] / ∂input[j] ∂input[k] = 0  ∀ i,j,k``."""
         raise NotImplementedError
 
-    def hessian_is_diagonal(self):
-        """Is `∂²output[i] / ∂input[j] ∂input[k]` nonzero only if `i = j = k`."""
+    def hessian_is_diagonal(self) -> bool:
+        """Is `∂²output[i] / ∂input[j] ∂input[k]` nonzero only if `i = j = k`.
+
+        The Hessian diagonal is only defined for layers that preserve the size
+        of their input.
+
+        Must be implemented by descendants that don't implement ``hessian_is_zero``.
+        """
         raise NotImplementedError
 
-    def hessian_diagonal(self):
+    # FIXME Currently returns `∂²output[i] / ∂input[i]² * g_out[0][i]`,
+    # which s the residual matrix diagonal, rather than the Hessian diagonal
+    def hessian_diagonal(
+        self, module: Module, g_in: Tuple[Tensor], g_out: Tuple[Tensor]
+    ) -> Tensor:
         """Return `∂²output[i] / ∂input[i]²`.
 
         Only required if `hessian_is_diagonal` returns `True`.
+
+        Args:
+            module: Module whose output-input Hessian diagonal is computed.
+            g_in: Gradients w.r.t. the module input.
+            g_out: Gradients w.r.t. the module output.
+
+        Returns:
+            Hessian diagonal. Has same shape as module input.
         """
         raise NotImplementedError
```

Comment:
- Move to `ElementwiseDerivatives` interface
- Create `residual_diagonal` function in `ElementwiseDerivatives` that calls out to `hessian_diagonal`
- Replace `hessian_diagonal` calls in extensions/core by `residual_diagonal`